lib/src/stream_provider.dart

```diff
@@ -6,9 +6,30 @@ import 'package:provider/src/provider.dart';
 
 typedef ErrorBuilder<T> = T Function(BuildContext context, Object error);
 
+/// Listens to a [Stream<T>] and exposes [T] to its descendants.
+///
+/// It is considered an error to pass a stream that can emit errors without providing
+/// a [orElse] method.
+///
+/// {@template provider.streamprovider.initialdata}
+/// [initialData] determines the value exposed until the [Stream] emits a value.
+/// If omitted, defaults to `null`.
+/// {@endtemplate}
+///
+/// {@macro provider.updateshouldnotify}
+///
+/// See also:
+///   * [Stream]
+///   * [StreamController], to create a [Stream]
 class StreamProvider<T>
     extends AdaptiveBuilderWidget<Stream<T>, StreamController<T>>
     implements SingleChildCloneableWidget {
+  /// Creates a [StreamController] from [builder] and subscribes to it.
```

Comment:
Just wondering why this doesn't require the user to return a `Stream<T>`? Feels weird to have a `StreamProvider` require you return a `StreamController`. I figured the builder would be a function that returns a `Stream`?

Reply:
It's similar to how `builder` from `ValueListenableProvider` returns a `ValueNotifier` instead of `ValueListenable`.

That's to free the resources with `.close()`/`.dispose()`.

If we have a 
```dart
StreamProvider(
  builder: (_) => StreamController().stream,
  ...
)
```
Then even with a `dispose` callback, there's no way to close the controller since its reference is lost.

Reply:
Not sure I'm convinced by this... Why would I need to create a StreamController inside the `builder`?

This is kinda how I thought the Widget would work when I first saw the name:

```
StreamProvider<QuerySnapshot>(
  builder: (_) => Firestore.instance.collection('books').snapshots(),
  ...
)
```

or

```
StreamProvider<List<Todo>>(
  builder: (_) => myBloc.todos,
  ...
)
```

or

```
StreamBuilder<String>(
  builder: (_) async* {
    yield "Loading"
    yield await fetchSomething();
  },
  ...
)
```

These would all be pretty awkward with the current API. You'd have to do something like this with all of them:

```
StreamProvider<QuerySnapshot>(
  builder: (_) => StreamController()..addStream(Firestore.instance.collection('books').snapshots()),
  ...
)
```

Overall, I get the need to release resources, but when it comes to Streams, that means cancelling a subscription, not disposing of a controller?

Reply:
In the examples you gave you'd use the value constructor:

```dart
StreamProvider.value(
  stream: Bloc.of(context).foo,
 ...
) 
```

Which takes a stream directly.


Canceling subscription/removing listeners is needed for both constructors. But only the builder constructor requires to call `close`/`dispose`

Reply:
Hrm, I thought the `builder` would be most useful to help peeps from recreating Streams over and over in the build method (the Firestore example would suffer from this using the `value` factory), rather than as a way to create your own `StreamController` that would be disposed.

Reply:
Hum, I get that.
At the same time I wanted to provide the most harmless solution possible, even if it limits the possibilities. 
I don't want folks to start misusing it and have memory leaks all over the place.


I mean with your suggestion we can technically do:

```dart
StreamBuilder(
  builder: (_) async* {
    var controller = StreamController();
    try {
      yield* controller.stream;
    } finally {
      controller.close();
    } 
  }
)
```

But that's pretty advanced

Reply:
For now I'll merge it. I'll open an issue to see if we can improve that aspect